helix-term/src/commands.rs

```diff
@@ -4533,18 +4535,122 @@ pub fn completion(cx: &mut Context) {
 }
 
 // comments
-fn toggle_comments(cx: &mut Context) {
+type CommentTransactionFn = fn(
+    token: Option<&str>,
+    tokens: Option<Vec<BlockCommentToken>>,
+    doc: &Rope,
+    selection: &Selection,
+) -> Transaction;
+
+fn toggle_comments_impl(cx: &mut Context, comment_transaction: CommentTransactionFn) {
     let (view, doc) = current!(cx.editor);
-    let token = doc
+    let token: Option<&str> = doc
+        .language_config()
+        .and_then(|lc| lc.comment_tokens.as_ref())
+        .map(|tc| tc.first())
+        .unwrap_or(None)
+        .map(|tc| tc.as_str());
+    let tokens: Option<Vec<BlockCommentToken>> = doc
         .language_config()
-        .and_then(|lc| lc.comment_token.as_ref())
-        .map(|tc| tc.as_ref());
-    let transaction = comment::toggle_line_comments(doc.text(), doc.selection(view.id), token);
+        .and_then(|lc| lc.block_comment_tokens.as_ref().cloned());
+
+    let transaction = comment_transaction(token, tokens, doc.text(), doc.selection(view.id));
 
     doc.apply(&transaction, view.id);
     exit_select_mode(cx);
 }
 
+fn toggle_comments(cx: &mut Context) {
+    toggle_comments_impl(cx, |token, tokens, doc, selection| {
+        let text = doc.slice(..);
+
+        // only have line comment tokens
+        if matches!((token, tokens.clone()), (Some(_), None)) {
+            return comment::toggle_line_comments(doc, selection, token);
+        }
+
+        let split_lines = comment::split_lines_of_selection(text, selection);
+
+        let (line_commented, line_comment_changes) = comment::find_block_comments(
+            tokens
+                .clone()
+                .unwrap_or_else(|| vec![BlockCommentToken::default()]),
```

Comment:
Here and in the `comment` module we can remove most of these clones by passing around slices of the Vec, so `&[BlockCommentToken]` rather than `Vec<BlockCommentToken>`. `comment::find_block_comments` will need to switch the line `let mut tokens = tokens` to `let mut tokens = tokens.to_vec()` which will allocate a new vec but we can get away with not allocating the vec otherwise.

Instead of `unwrap_or_else` in this file we will need:

```rust
let default_block_comment_tokens = &[BlockCommentToken::default()];
let block_comment_tokens = tokens.unwrap_or(default_block_comment_tokens);
```

so that the slice lives long enough